### PR TITLE
add ci for karmada operator

### PR DESCRIPTION
**Pull request by zhzhuang-zju**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
The current `karmada-operator` lacks code merge access control; this PR will introduce:
- A one-click script to install a Karmada instance through the `karmada-operator`.
- karmada-operator ci

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
```

### CODE REVIEW

**Comment by zhzhuang-zju** [comment]
cc @RainbowMango @jabellard

**Comment by zhzhuang-zju** [comment]
test report:

```yaml
# karmada.yaml
apiVersion: operator.karmada.io/v1alpha1
kind: Karmada
metadata:
  name: karmada-demo
  namespace: test
spec:
  crdTarball:
    httpSource:
      url: https://github.com/karmada-io/karmada/releases/download/v1.11.0/crds.tar.gz
  components:
    etcd:
      local:
        imageRepository: registry.k8s.io/etcd
        imageTag: 3.5.13-0
        replicas: 1
        volumeData:
          # hostPath:
          #   type: DirectoryOrCreate
          #   path: /var/lib/karmada/etcd/karmada-demo
          volumeClaim:
            metadata:
              name: etcd-data
            spec:
              accessModes:
                - ReadWriteOnce
              resources:
                requests:
                  storage: 3Gi
    karmadaAPIServer:
      imageRepository: registry.k8s.io/kube-apiserver
      imageTag: v1.29.6
      replicas: 1
      serviceType: NodePort
      serviceSubnet: 10.96.0.0/12
    karmadaAggregatedAPIServer:
      imageRepository: docker.io/karmada/karmada-aggregated-apiserver
      imageTag: v1.11.0
      replicas: 1
    karmadaControllerManager:
      imageRepository: docker.io/karmada/karmada-controller-manager
      imageTag: v1.11.0
      replicas: 1
    karmadaScheduler:
      imageRepository: docker.io/karmada/karmada-scheduler
      imageTag: v1.11.0
      replicas: 1
    karmadaWebhook:
      imageRepository: docker.io/karmada/karmada-webhook
      imageTag: v1.11.0
      replicas: 1
    kubeControllerManager:
      imageRepository: registry.k8s.io/kube-controller-manager
      imageTag: v1.29.6
      replicas: 1
    karmadaMetricsAdapter:
      imageRepository: docker.io/karmada/karmada-metrics-adapter
      imageTag: v1.11.0
      replicas: 2
    # karmadaSearch: # the component `Karmadasearch` is not installed by default, if you need to install it, uncomment it and note the formatting
      # imageRepository: docker.io/karmada/karmada-search
      # imageTag: v1.8.0
      # replicas: 1
    karmadaDescheduler: # the component `KarmadaDescheduler` is not installed by default, if you need to install it, uncomment it and note the formatting
      imageRepository: docker.io/karmada/karmada-descheduler
      imageTag: v1.11.0
      replicas: 1
  hostCluster:
    networking:
      dnsDomain: cluster.local
```
```
$ kubectl create secret generic my-kubeconfig --from-file=config=$HOME/.kube/karmada.config --namespace karmada-system 
$ kubectl apply -f operator/config/deploy/karmada-operator.yaml
$ kubectl apply -f operator/config/crds/
$ kubectl create namespace test
$ kubectl apply -f karmada.yaml
$ kubectl get pods -ntest                                                                 
NAME                                                    READY   STATUS    RESTARTS   AGE
karmada-demo-aggregated-apiserver-7c94956588-kq6m6      1/1     Running   0          49m
karmada-demo-apiserver-9cf845cf4-bkvhk                  1/1     Running   0          49m
karmada-demo-controller-manager-55fb77c55-r4mtr         1/1     Running   0          48m
karmada-demo-descheduler-5b6465697d-xbbqh               1/1     Running   0          48m
karmada-demo-etcd-0                                     1/1     Running   0          49m
karmada-demo-kube-controller-manager-6f8f9b6cf8-zmmsg   1/1     Running   0          48m
karmada-demo-metrics-adapter-686dbd65d4-5tltz           1/1     Running   0          48m
karmada-demo-metrics-adapter-686dbd65d4-v67dd           1/1     Running   0          48m
karmada-demo-scheduler-7d5846d6db-nspv4                 1/1     Running   0          48m
karmada-demo-webhook-5bf69bc674-nh7cp                   1/1     Running   0          48m
```

**Comment by zhzhuang-zju** [comment]
cc @chaosi-zju

**Comment by RainbowMango** [comment]
@zhzhuang-zju How to reproduce it? Does #5040 introduce this issue?

**Comment by zhzhuang-zju** [comment]
> @zhzhuang-zju How to reproduce it? Does #5040 introduce this issue?

Yes, Just follow the steps as described in comments https://github.com/karmada-io/karmada/pull/5519#issuecomment-2347960323 and will reproduce this issue. I think we can add a CI for installation by the `karmada-operator`, like #4401

**Comment by RainbowMango** [comment]
> I think we can add a CI for installation by the karmada-operator, like https://github.com/karmada-io/karmada/pull/4401

We do need CI, but a simple CI like #4401 cannot cover all cases, we still need E2E cases to cover features like #5242, #5272, #5448.

I will take a look at #4401 first.

**Comment by zhzhuang-zju** [comment]
> I will take a look at https://github.com/karmada-io/karmada/pull/4401 first.

Recently, karmada-operator has added many new features. The installation verification in #4401 might not be sufficient anymore. For example, it does not validate the installation of the component `karmada-scheduler-estimator`. Collaborating with #5511 could help resolve this issue, so #4401 needs to be updated.

**Comment by RainbowMango** [comment]
OK, please cc me again #4401 once it's ready.

By the way, I think this PR can help update https://github.com/karmada-io/karmada/blob/master/operator/config/samples/karmada.yaml, so that the bugs can be found more easily.

**Comment by RainbowMango** [comment]
@zhzhuang-zju Please refresh this PR, I see #4401 closed.

File: hack/deploy-karmada-operator.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+kubectl --kubeconfig="${KUBECONFIG}" --context="${CONTEXT_NAME}" apply -f "${REPO_ROOT}/artifacts/deploy/namespace.yaml"
+
+# create secret my-kubeconfig
+kubectl --kubeconfig="${KUBECONFIG}" --context="${CONTEXT_NAME}" create secret generic my-kubeconfig --from-file=config="${KUBECONFIG}" --namespace ${KARMADA_SYSTEM_NAMESPACE}
+
+# deploy karmada-operator
+kubectl --kubeconfig="${KUBECONFIG}" --context="${CONTEXT_NAME}" apply -f "${REPO_ROOT}/operator/config/deploy/karmada-operator.yaml"
+
+# install Karmada operator crds
+kubectl --kubeconfig="${KUBECONFIG}" --context="${CONTEXT_NAME}" apply -f operator/config/crds/
```
Comment (RainbowMango) [inline hack/deploy-karmada-operator.sh:70]: I think the CRD should be applied ahead of the operator to avoid unexpected warnings or restart attempts.

**Comment by zhzhuang-zju** [comment]
This PR introduces three new scripts:
- hack/deploy-karmada-operator.sh: deploys karmada-operator on the specified cluster
- hack/deploy-karmada-by-operator.sh: deploys karmada instance to a given cluster via karmada-operator.
- hack/local-up-karmada-by-operator.sh: Combining the two scripts above to provide the capability for a one-click installation of a Karmada instance.

Refer to https://github.com/karmada-io/karmada/actions/runs/11270900302/job/31342651001?pr=5519 and local verifiation, using `hack/local-up-karmada-by-operator.sh` can create a local Karmada environment with basic functionality working properly.
```
$ karmadactl get cluster
NAME      CLUSTER   VERSION   MODE   READY   AGE    ADOPTION
member1   Karmada   v1.31.0   Push   True    104m   -
member2   Karmada   v1.31.0   Push   True    104m   -
member3   Karmada   v1.31.0   Pull   True    104m   -
$ karmadactl get deployments --operation-scope all
NAME    CLUSTER   READY   UP-TO-DATE   AVAILABLE   AGE   ADOPTION
nginx   Karmada   2/2     2            2           68m   -
nginx   member1   1/1     1            1           67m   Y
nginx   member2   1/1     1            1           67m   Y
```
However, to fully validate the functionality through e2e tests, it is necessary to wait for the merge of the following PRs:
- https://github.com/karmada-io/karmada/pull/5511 support deploy estimator in the specified namespace
- https://github.com/karmada-io/karmada/pull/5573 avoid installation failures for 'karmada-scheduler-estimator' and 'karmada-interpreter-webhook-example' due to non-uniform secret names.

Therefore, I suggest that the current Operator CI be used to verify the capability of the Karmada Operator to deploy Karmada instances and basic functionalities, such as the successful execution of `karmadactl join`. E2E tests as well as other tests can be introduced later.

wdyt? @liangyuanpeng @RainbowMango

File: .github/workflows/operator_test.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          tool-cache: false
+          # all of these default to true, but feel free to set to "false" if necessary for your workflow
+          android: true
+          dotnet: true
+          haskell: true
+          large-packages: false
+          docker-images: false
+          swap-storage: false
+      - name: checkout code
+        uses: actions/checkout@v4
```
Comment (RainbowMango) [inline .github/workflows/operator_test.yaml:45]: We need to build the operator from the source code, right?

File: .github/workflows/operator_test.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    runs-on: ubuntu-22.04
+    strategy:
+      fail-fast: false
+      matrix:
+        # Here support the latest three minor releases of Kubernetes, this can be considered to be roughly
+        # the same as the End of Life of the Kubernetes release: https://kubernetes.io/releases/
+        # Please remember to update the CI Schedule Workflow when we add a new version.
+        k8s: [ v1.29.0, v1.30.0, v1.31.0 ]
+    # prevent job running from forked repository
+    if: ${{ github.repository == 'karmada-io/karmada' }}
```
Comment (RainbowMango) [inline .github/workflows/operator_test.yaml:29]: Why?

File: hack/local-up-karmada-by-operator.sh [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+# This script is used in workflow to validate karmada installation by operator.
+# It starts a local karmada control plane based on current codebase via karmada-operator and with a certain number of clusters joined.
+# This script depends on utils in: ${REPO_ROOT}/hack/util.sh
+# 1. used by developer to setup develop environment quickly.
+# 2. used by e2e testing to test if the operator installs correctly.
+
+function usage() {
+    echo "Usage:"
+    echo "    hack/verify-operator-installed.sh [-h]"
```
Comment (RainbowMango) [inline hack/local-up-karmada-by-operator.sh:28]:
```suggestion
    echo "    hack/local-up-karmada-by-operator.sh [-h]"
```

File: hack/util.sh [hunk 5]
```diff
@@ -257,6 +257,19 @@ function util::append_client_kubeconfig {
     kubectl config set-context "${client_id}" --cluster="${client_id}" --user="${client_id}" --kubeconfig="${kubeconfig_path}"
 }
 
+# util::append_client_kubeconfig_with_server creates a new context including a cluster and a user to the existed kubeconfig file
+function util::append_client_kubeconfig_with_server {
```
Comment (RainbowMango) [inline hack/util.sh:261]: Seems the function name not that straightforward.

File: .github/workflows/operator_test.yaml (hunk 2 above)
Comment (zhzhuang-zju) [reply .github/workflows/operator_test.yaml:45]: yes

File: .github/workflows/operator_test.yaml (hunk 3 above)
Comment (zhzhuang-zju) [reply .github/workflows/operator_test.yaml:29]: Good point! We should remove the restriction on the forked repository.

File: hack/util.sh (hunk 5 above)
Comment (zhzhuang-zju) [reply hack/util.sh:261]: How about `append_existed_kubeconfig_with_new_context_name`?

**Comment by zhzhuang-zju** [comment]
Since #5586 modified the karmada-operator manifest and added files such as clusterrole, this PR needs to wait for #5586 to be merged first.

/hold

File: .github/workflows/operator_test.yaml [hunk 6]
```diff
@@ -0,0 +1,82 @@
+name: Operator Test
```
Comment (RainbowMango) [inline .github/workflows/operator_test.yaml:1]:
```suggestion
name: Operator
```

File: .github/workflows/operator_test.yaml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  pull_request:
+# https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency
+concurrency:
+  group: ${{ github.workflow }}-${{ github.actor }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+permissions:
+  contents: read  # for actions/checkout to fetch code
+jobs:
+  set-up-operator-test-environment:
+    name: Set up operator test environment
```
Comment (RainbowMango) [inline .github/workflows/operator_test.yaml:19]:
```suggestion
  test-on-kubernetes-matrix:
    name: test on Kubernetes
```

Comment (RainbowMango) [reply .github/workflows/operator_test.yaml:19]: The current test name on PR is:
```
Operator Test / Set up operator test environment (v1.29.0) (pull_request) 
```
Suggest name:
```
Operator / Test on Kubernetes (v1.29.0) (pull_request) 
```

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**